File: rdfbean-core/src/main/java/com/mysema/rdfbean/model/QNODE.java
package com.mysema.rdfbean.model;

import javax.annotation.Nullable;

import com.mysema.query.types.Order;
import com.mysema.query.types.OrderSpecifier;
import com.mysema.query.types.expr.Param;

/**
 * @author tiwe
 *
 * @param <T>
 */
public class QNODE<T extends NODE> extends Param<T>{

    private static final long serialVersionUID = 1134119241723346776L;

    public static final QID s = new QID("s");

    public static final QUID p = new QUID("p");

    public static final QNODE<NODE> o = new QNODE<NODE>(NODE.class,"o");

    public static final QID c = new QID("c");

    public static final QID type = new QID("type");

    public static final QID typeContext = new QID("typeContext");

    public static final QNODE<NODE> first = new QNODE<NODE>(NODE.class, "first");

    public static final QID rest = new QID("rest");

    @Nullable
    private volatile QLIT lit;

    @Nullable
    private volatile QID id;

    @Nullable
    @SuppressWarnings("unchecked")
    private volatile OrderSpecifier asc, desc;

    public QNODE(Class<T> type, String variable) {
        super(type, variable);
    }

    public PatternBlock is(Object predicate, Object subject){
        return Blocks.pattern(subject, predicate, this);
    }

    @SuppressWarnings("unchecked")
    public OrderSpecifier asc(){
        if (asc == null){
            asc = new OrderSpecifier(Order.ASC, this);
        }
        return asc;
    }

    @SuppressWarnings("unchecked")
    public OrderSpecifier desc(){
        if (desc == null){
            desc = new OrderSpecifier(Order.DESC, this);
        }
        return desc;
    }

    public QID id(){
        if (id == null){
            id = new QID(getName());
        }
        return id;
    }

    public QLIT lit(){
        if (lit == null){
            lit = new QLIT(getName());
        }
        return lit;
    }

    @Override
    public boolean equals(Object o) {
        if (o == this){
            return true;
        }else if (o instanceof QNODE<?>){
            QNODE<?> other = (QNODE<?>)o;
            return other.getName().equals(getName());
        }else{
            return false;
        }
    }

    @Override
    public int hashCode(){
        return getName().hashCode();
    }

}

File: rdfbean-scala/src/main/scala/com/mysema/rdfbean/scala/QNODE.scala
package com.mysema.rdfbean.scala;

import java.util.Collection
import javax.annotation.Nullable
import com.mysema.query.types.{ Expression, ConstantImpl, Ops, Order, OrderSpecifier, ParamExpression, Visitor }
import com.mysema.query.scala._
import com.mysema.rdfbean.model.NODE
import com.mysema.query.scala.{ Operations, SimpleExpression }
import com.mysema.rdfbean.model.{ Blocks, NODE, ID, LIT, BID, UID, RDF }

//import scala.collection.JavaConversions._

object QNODE {

  val s = new QID("s")

  val p = new QID("p")

  val o = new QNODE[NODE](classOf[NODE], "o")

  val c = new QID("c")

  val t = new QID("type")

  val typeContext = new QID("typeContext")

  val first = new QNODE[NODE](classOf[NODE], "first")

  val rest = new QID("rest")

}

class QNODE[T <: NODE](val t: Class[T], val name: String) extends ParamExpression[T] with SimpleExpression[T] {

  def accept[R, C](v: Visitor[R, C], context: C): R = v.visit(this, context)

  def getName() = name

  def getType() = t

  def isAnon() = false

  def getNotSetMessage() = "A parameter of type " + getType.getName + " was not set"

  def is(p: AnyRef, s: AnyRef) = Blocks.pattern(s, p, this)

  lazy val id = new QID(getName)

  lazy val lit = new QLIT(getName)

  lazy val asc = new OrderSpecifier[String](Order.ASC, this.asInstanceOf[Expression[String]])

  lazy val desc = new OrderSpecifier[String](Order.DESC, this.asInstanceOf[Expression[String]])

  lazy val stringValue = Operations.string(Ops.STRING_CAST, this)

  override def equals(o: Any): Boolean = {
    o match {
      case n: QNODE[_] => n.getName == name
      case _ => false
    }
  }

  override def hashCode = getName.hashCode

}

